grakel_replace/torch_wl_kernel.py

```diff
@@ -162,85 +248,75 @@ def __init__(self, n_iter: int = 5, *, normalize: bool = True) -> None:
+        """Extract neighborhood information from adjacency matrix."""
         if adj.layout == torch.sparse_csr:
             adj = adj.to_sparse_coo()
 
         adj = adj.coalesce()
-        indices = adj.indices()
-        rows, cols = indices
-        num_nodes = labels.size(0)
+        rows, cols = adj.indices()
+        num_nodes = adj.size(0)
 
-        # Create a mask for each node's neighbors
-        neighbor_mask = torch.zeros((num_nodes, num_nodes), dtype=torch.bool,
-                                    device=self.device)
-        neighbor_mask[rows, cols] = True
+        neighbors = [[] for _ in range(num_nodes)]
+        for row, col in zip(rows.tolist(), cols.tolist(), strict=False):
+            neighbors[row].append(col)
 
-        # Get neighbor labels for each node
-        neighbor_labels = labels.unsqueeze(0).expand(num_nodes, -1)
-        neighbor_labels = neighbor_labels.masked_fill(~neighbor_mask, -1)
+        return neighbors
 
-        # Sort neighbor labels for each node
-        sorted_neighbor_labels, _ = torch.sort(neighbor_labels, dim=1, descending=True)
+    def _wl_iteration(self, adj: Tensor, labels: Tensor) -> Tensor:
+        """Perform one WL iteration exactly matching GraKel's implementation."""
+        num_nodes = labels.size(0)
+        new_labels = []
+
+        neighbors = self._get_node_neighbors(adj)
 
-        # Remove padding (-1 values) from sorted labels
-        valid_neighbors_mask = sorted_neighbor_labels != -1
-        max_neighbors = valid_neighbors_mask.sum(1).max().item()
-        sorted_neighbor_labels = sorted_neighbor_labels[:, :max_neighbors]
+        for node_idx in range(num_nodes):
+            # Get current node label
+            node_label = labels[node_idx].item()
 
-        # Combine node labels with neighbor labels
-        node_labels_expanded = labels.unsqueeze(1).expand(-1, max_neighbors)
+            # Get sorted neighbor labels - exactly as GraKel does
+            neighbor_labels = sorted([labels[n].item() for n in neighbors[node_idx]])
 
-        # Create feature vectors
-        features = torch.cat([
-            node_labels_expanded.unsqueeze(-1).float(),
-            sorted_neighbor_labels.unsqueeze(-1).float()
-        ], dim=-1)
+            # Create credential string exactly matching GraKel's format:
+            # "current_label,sorted_neighbor_labels"
+            credential = f"{node_label},{neighbor_labels}"
 
-        # Hash the combined features
-        hashed_features = self.hash_module(features).squeeze(-1)
-        hashed_labels = hashed_features.sum(dim=1)
+            # Update label dictionary
+            if credential not in self.label_dict:
+                self.label_dict[credential] = len(self.label_dict)
 
-        # Convert to discrete labels
-        _, new_labels = torch.unique(hashed_labels, sorted=True, return_inverse=True)
-        return new_labels
+            new_labels.append(self.label_dict[credential])
```

Comment:
Again, definitely not a bottleneck, and does not need to be put in, just reading through with my optimization cap on and sharing some micro-optimization techniques.

In the case where you go inside the `if` block, you already have the new label `len(self.label_dict)` and so you do not need to rehash and lookup in `self.label_dict`, one less dict lookup :)

```python
            # Update label dictionary
            if credential not in self.label_dict:
                label = len(self.label_dict)
                self.label_dict[credential] = label
            else:
                label = self.label_dict[credential]
            
            new_labels.append(label)
            
```

Reply:
There's also this method fyi, nifty one liner but arguable less readable
https://docs.python.org/3/library/stdtypes.html#dict.setdefault

```python
new_labels.append(self.label_dict.setdefault(credential, len(self.label_dict))
```